Cython/Compiler/ParseTreeTransforms.py

```diff
@@ -2545,6 +2545,26 @@ def visit_ClassDefNode(self, node):
         self.qualified_name = orig_qualified_name
         return node
 
+    def visit_NameNode(self, node):
+        scope = self.current_env()
+        if scope.is_c_class_scope:
+            # unlike for a PyClass scope, these attributes aren't defined in the
+            # dictionary when the class definition is executed, so they need
+            # replacing with compile-time constants
```

Comment:
What happens if users assign to these names inside of the class scope?

Reply:
I missed that. I guess it can be replaced with the nodes for something like `locals().get('__qualname__', CONSTANT_VALUE)`. I'll have a go at that on the next revision.

Reply:
Hmm. There's also `del`. Not sure, but we might really get away with declaring cdef class namespaces more special than Python class scopes, which really is the case in some regards.

Reply:
`del` turns out to be broken inside cdef class scopes anyway. I've rewritten it in a way that should behave right when in future though